README.md
# mobro-raspberrypi

![GitHub tag (latest by date)](https://img.shields.io/github/v/tag/ModBros/mobro-raspberrypi?label=version)
![GitHub](https://img.shields.io/github/license/ModBros/mobro-raspberrypi)
[![Discord](https://img.shields.io/discord/620204412706750466?label=Discord)](https://discord.com/invite/DSNX4ds)
[![FAQ](https://img.shields.io/badge/-FAQ-f30.svg)](https://www.mod-bros.com/en/faq/mobro)
[![YouTube](https://img.shields.io/badge/-YouTube-red.svg)](https://www.youtube.com/channel/UCUU5DVHRzQEnGPVS5WKRg6A)
[![Download](https://img.shields.io/badge/-Download-brightgreen.svg)](https://www.mod-bros.com/en/projects/mobro)

**Official Raspberry Pi image of MoBro**

![MoBro logo + header](./images/readme_header.png)

## Table of Contents

* [What is MoBro?](#what-is-mobro)
* [Windows application](#windows-application)
* [Raspberry Pi image](#raspberry-pi-image)
  * [What does this image do?](#what-does-this-image-do)
  * [Download and install](#download-and-install)
  * [Supported Raspberry Pi Models](#supported-raspberry-pi-models)
* [Technical documentation](#technical-documentation)
  * [File system](#file-system)
  * [Configuration](#configuration)
  * [Custom modifications](#custom-modifications)
  * [REST Api](#rest-api)
  * [Logging](#logging)

Feel free to visit our [Forum](https://www.mod-bros.com/en/forum)
or [Discord](https://discord.com/invite/DSNX4ds) if you run into issues or questions that are not already answered here.  
Feedback as well as suggestions for additional features and improvements are always welcome!

# What is MoBro?

The Monitor Bro (MoBro) by ModBros collects monitoring data about your installed hardware locally on your PC.  
It is designed and built to take in data from multiple different monitoring applications
(= data sources or plugins) such as HWiNFO and others.  
It combines them into a single UI while letting you choose which values of which source you are interested in and want
to see.  
All configurable via an easy customization interface.

__Defining features__:

* Reading data from different sources
* Customizable interface displaying the data YOU are interested in
* Your data does not leave your network and is shared with no one
* Displaying data on various client devices (Raspberry Pi, Android phones,...) located anywhere in your house
* Different customizable theme for each of your devices

# Windows application

The running MoBro PC application is required for this Raspberry Pi project.  
Currently only available for Windows ([Download here](https://www.mod-bros.com/en/projects/mobro)).

[![Windows app explanation video](http://img.youtube.com/vi/bLwOTQ8MW7s/0.jpg)](http://www.youtube.com/watch?v=bLwOTQ8MW7s)


# Raspberry Pi image

This Raspberry Pi image acts as a client device to the MoBro Windows application to which it connects.  
It provides an easy and cost-effective way to set up a wireless device displaying the PC's stats in realtime anywhere in
the house.

[![Installation video](http://img.youtube.com/vi/iebBcQuBhYs/0.jpg)](http://www.youtube.com/watch?v=iebBcQuBhYs)

## What does this image do?

This custom pre-configured image provides an easy way to set up the Raspberry Pi as a MoBro monitoring device.  
It is ready to be flashed onto a micro SD card and put straight to use in a Raspberry Pi.   
All the required configuration is done via an easy-to-use web based configuration wizard.

No coding skills or Linux experience required.

## Download and install

Detailed instructions on how to download flash and setup the image can be found here:  
[Download and flash](https://www.mod-bros.com/en/faq/mobro/raspberry/download),
[Setup](https://www.mod-bros.com/en/faq/mobro/raspberry/setup)

## Supported Raspberry Pi Models

This image is ready to run on all Raspberry Pi models.  
For wireless operation a model with built-in Wifi is required.

[Supported models and known limitations](https://www.mod-bros.com/en/faq/mobro/raspberry/supported-hardware)


# Technical documentation

This image is based on the official [Raspberry Pi OS Lite](https://www.raspberrypi.org/software/operating-systems/).

Changelog can be found here: [Changelog](./CHANGELOG.md)  
Setup guide for this repository: [Setup](./SETUP.md)

The mobro service ([mobro.sh](./service/mobro.sh)) is the centerpiece of this image and handles all the automation and
heavy lifting.  
User configuration is done via a simple web based wizard hosted by Lighttpd.

## File system

The disk of this image is split up into three partitions: / (root), /boot and /mobro

### root (OverlayFS)

The __root__ partition is set up for [OverlayFS](https://en.wikipedia.org/wiki/OverlayFS). So the underlying partition
is mounted read-only with an in memory read/write partition overlay on top if it.  
As a consequence, all file changes will only be stored in the overlay in RAM and not actually written
to the SD card and therefore will be gone after the next shutdown/reboot.  

This has the following benefits:
* no write operations to the SD card, which greatly extends its lifespan
* reduced risk of data corruption in case the Pi is not shut down correctly (e.g. just cutting the power)
* guarantee that the Pi is in the exact same state on every boot

However, this approach also complicates the configuration process as well as all custom modifications to the image.  
See the section below ([Custom modifications](#custom-modifications)).

In case you wish to disable OverlayFS altogether you can do this from the 'Advanced customization' step in the 
configuration wizard.

### /boot + /mobro
The __/boot__ partition is mounted read-only per default. In case you need to manually alter e.g. the config.txt file you
first need to remount /boot with write permissions.  
Simply call `status` to see the current mounting status and
`set_boot_configurable` to remount the /boot partition with write permissions.

__/mobro__ is a small dedicated partition that is mounted with write permissions and is used to store configuration
files, flags etc. required by the mobro service.

## Custom modifications 

As a consequence of using OverlayFS, all custom modifications to configuration files on the root partition will be lost 
upon reboot or shutdown. In order to persist custom changes you will need to:
1. disable OverlayFS + temporarily disable the mobro service  
   (the service would automatically re-enable OverlayFS and reboot before we can do anything if we don't disable it)
   ```
   set_root_configurable
   ```
   Note: the Pi will reboot after this
2. re-mount /boot with write permissions
   ```
   set_boot_configurable
   ```
3. apply all your custom changes
4. enabling OverlayFS + mobro service again
   ```
   set_default
   ```
   Note: This might take a minute. The Pi will again reboot after this

Note:  
The procedure explained above is only required if you need to install software or apply custom changes that are
not covered by our configuration wizard. For all the changes done through our configuration wizard, the above scenario
is handled automatically by our scripts and requires no additional user handling or input. Just hist 'apply' as usual
and you're good ;)

In case you wish to disable OverlayFS altogether you can do this from the 'Advanced customization' step in the 
configuration wizard.


## Configuration

The MoBro specific configuration is stored as a simple property file.  
Current settings and their meaning:

| Setting                    | Description |
| :------------------------- | :---------- |
| localization_country       | [ISO 3166 country code](https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes) |
| localization_timezone      | timezone name as listed in [TZ database](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones) |
| network_mode               | 'eth', 'wifi' or 'usb' (Pi Zero only) |
| network_ssid               | the network SSID to connect to (only in wifi mode) |
| network_pw                 | the network password (only in wifi mode) |
| network_wpa                | the WPA mode. one of: 2a, 2t, 1t, n (optional, only in wifi) |
| network_hidden             | whether the network is hidden (only in wifi) |
| discovery_mode             | 'auto' (default) or 'manual'|
| discovery_key              | the discovery key as configured in the MoBro deskop application (default=mobro) |
| discovery_ip               | the static IP address of the PC (only in 'manual' network mode) |
| display_driver             | 'default', 'pi7' or path to the driver install executable |
| display_rotation           | one of: 0 (default), 90, 180, 270 |
| display_hdmi_mode          | the hdmi_group and hdmi_mode seperated by ':' (e.g. 2:4) |
| display_screensaver        | 'disabled', 'custom', screensaver file or URL |
| display_screensaver_url    | URL used as screensaver when mode set to 'custom' |
| display_delay              | delay for the screensaver in minutes, default=5 |
| advanced_overclock_mode    | overclock configuration to apply, default=none |
| advanced_overclock_consent | explicit consent to overclocking. has to be set to '1' for the overclock to be applied |
| advanced_fs_dis_overlayfs  | flag to disable the OverlayFS, default = 0 |
| advanced_fs_persist_log    | flag to enable persistent logging, default = 0 |

Just altering the values in the configuration file is NOT enough. The configuration file has to be applied by executing
the [apply_new_config.sh](./scripts/apply_new_config.sh) script and passing the config file as parameter.

Example configuration:

```
localization_country=AT
localization_timezone=UTC
network_mode=
network_ssid=
network_pw=
network_wpa=
network_hidden=0
discovery_mode=auto
discovery_key=mobro
discovery_ip=
display_driver=default
display_hdmi_mode=
display_rotation=0
display_screensaver=clock_date.php
display_delay=0
```

In addition, it is possible to pass a 2nd configuration file to [apply_new_config.sh](./scripts/apply_new_config.sh)
that contains custom entries that will be added to the config.txt file. E.g.:

```
temp_limit=75
arm_freq=1000
over_voltage=0
```

Order of application for changes to the config.txt file:

1. default config.txt file
2. display driver installation
3. overclock settings (advanced_overclock from MoBro config file)
4. custom overrides/additions (from the 2nd config file mentioned above)
5. OverlayFS settings

## REST Api

For debug purposes the MoBro Raspberry Pi image exposes a very basic REST API.  
This API is subject to change.

|     | Route              | Description                                                                | Format | Parameters |
| --- | :----------------- | :------------------------------------------------------------------------- | :----- | :--------- |
| GET | /api               | returns the API documentation                                              | text   | - |
| GET | /api/version       | returns the current version number                                         | text   | - |
| GET | /api/log           | returns the current logfile                                                | text   | *lines*: only most recent n lines<br> *count*: log files of previous n boots (max = 10, default = 0) |
| GET | /api/syslog        | returns the current syslog file                                            | text   | - |
| GET | /api/top           | returns the output of the "top" command. i.e.: CPU/RAM Usage, Processes... | text   | - |
| GET | /api/configuration | returns the current configuration                                          | text   | - |
| GET | /api/hwstats       | stats of the hardware components (e.g. CPU/RAM usage, temperature,...)     | json   | *filter*: cpu,memory,filesystem,wifi (default=none) |
| PUT | /api/restart       | restarts the Raspberry Pi                                                  | -      | *delay*: minutes before restart (default: 0) |
| PUT | /api/shutdown      | shuts down the Raspberry Pi                                                | -      | *delay*: minutes before shutdown (default: 0) |
| PUT | /api/service       | starts, stops or restarts the MoBro service                                | -      | *action*: start, stop, restart (default) |


## Logging

This image uses in-memory busybox-syslogd. The current log file can be viewed by calling 'logread' or through the
respective ['syslog' REST endpoint](#rest-api).

The mobro specific log file containing the log output of all our services and scripts is located at __/tmp/mobro_log__
and can also be pulled through the respective ['log' REST endpoint](#rest-api).

The log file is not written to the SD-card and therefore not persisted. To persist log files for debug purposes, the
[shutdownlog](./service/shutdownlog.sh) service can be enabled in the configuration wizard. It will automatically copy
the current log file over to '/mobro/log' before regular shutdowns and rotate them to keep the most recent 10 files.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)

## [v12.3](https://github.com/ModBros/mobro-raspberrypi/compare/v12.2...v12.3) - 2021-11-17

### Added
* support for the new Raspberry Pi Zero 2 W
* resolution selection for displays connected over HDMI

### Changed
* updated included display driver repositories
* replaced deprecated 'display_rotate' with 'display_hdmi_rotate' and 'display_lcd_rotate' 

## [v12.2](https://github.com/ModBros/mobro-raspberrypi/compare/v12.1...v12.2) - 2021-08-10

### Changed
* Significantly reduced time until screensaver is shown after booting up
* Added modbros user to video group

### Fixed
* usage of Wi-Fi SSIDs containing spaces
* application of custom config.txt entries

## [v12.1](https://github.com/ModBros/mobro-raspberrypi/compare/v12...v12.1) - 2021-07-12

### Added

* allow custom Url as screensaver
* new additional screensavers

### Fixed
* display rotation for some GoodTFT and Waveshare drivers
* ethernet connection problems

## [v12](https://github.com/ModBros/mobro-raspberrypi/compare/v11.1...v12) - 2021-06-20

### Added
* syslog debug endpoint
* log system throttled state
* cleartext toggle for wifi password field
* MoBro splashscreen instead of console boot log
* rest api endpoint for hardware stats (cpu, memory, filesystem, wifi)
* OverlayFS to avoid SD card corruption
* overclocking of Pi 1, 2 and Zero from configuration wizard
* custom additions or overrides of config.txt from configuration wizard
* support for USB gadget mode for Pi Zero

### Changed
* switched to busybox in-memory logger
* wait for reconnect on wifi connection loss instead of opening configuration hotspot again
* minor styling adjustments

### Fixed
* wifi check on first boot
* Raspberry Pi acting as DHCP server in connected wifi under rare condition


## [v11.1](https://github.com/ModBros/mobro-raspberrypi/compare/v11...v11.1) - 2020-08-31

### Fixed
* default configuration for HDMI screens  
(wrong resolution of 640x480 was set for all screens)

## [v11](https://github.com/ModBros/mobro-raspberrypi/compare/v10...v11) - 2020-08-29

### Added

* added readme
* added changelog
* support for display rotation
* flags in country selection
* configurable screensaver
* timezone configuration
* added API calls for configuration and service 

### Changed

* reduced recurring background logging
* multiple connection attempts in background check  
(to avoid disconnection after just one bad request)
* better styling for select items
* merged MoBro specific configurations into a single file

### Fixed

* Pi didn't reconnect to PC after it lost connection once 


## [v10](https://github.com/ModBros/mobro-raspberrypi/compare/v9...v10) - 2020-05-21

### Fixed

* fixed connection over static IP
* multiple connections to the PC application were opened in some cases 

## [v9](https://github.com/ModBros/mobro-raspberrypi/compare/v8...v9) - 2020-05-17

### Added

* all new configuration wizard
* enabled display driver installation via web configuration
* enable configuration of wifi country (needed for correct 5G frequency bands)
* support for hidden wifi networks
* expose additional wifi settings via configuration (WPA version, encryption method)
* support for connection via static IP instead of service discovery
* set 'noatime' for /var/run, /var/log, /boot and /  
(to avoid writing to SD card every time a file is read) 
* added display resolution as parameter to mobro url
* added device type to mobro url
* added license

### Changed

* major re-design of the configuration page
* disabled swapping
* mount /var/run and /var/log as tmpfs 
* updated _config.txt_
* major project restructuring
* renamed service
* adapted parameters for log endpoint
* styling adjustments

### Fixed

* fixed hotspot creation in case no display is connected
* only show distinct SSIDs in configuration
* set uuid (=mac address) based on current network mode


## [v8](https://github.com/ModBros/mobro-raspberrypi/compare/v7...8) - 2020-04-19

### Changed

* removed Wifi password from configuration hotspot
* updated _config.txt_
* cleanup of dependencies

### Fixed

* unblock the Wifi interface on startup
* fixed CPU frequency bug on Pi 4

## [v7](https://github.com/ModBros/mobro-raspberrypi/compare/v7...v6) - 2020-03-29

### Added

* restful api endpoint to shutdown or restart the Pi 
* restful api endpoint to retrieve load data (top)
* restful api endpoint to retrieve current Pi temperature
* new dependencies: _xserver-xorg-video-fbturbo, git, secure-delete_

### Fixed

* removed chromium _use-gl_ flag as it caused issues with GPIO displays
* re-added lighttpd to autostart to make log available if service fails to start

## [v6](https://github.com/ModBros/mobro-raspberrypi/compare/v6...v5) - 2020-03-17

### Added

* added LAN support
* expose version via rest endpoint
* included waveshare display drivers on image
* added uuid to mobro url

### Changed

* adapted _config.txt_
* disable screen turning off
* re-branded to use MoBro logo
* update to log rest endpoint
* code refactoring and project restructuring


### Fixed

* fixed stopping of processes
* workaround for chromium update bug
* fixed bug on changing the discovery key


## [v5](https://github.com/ModBros/mobro-raspberrypi/compare/v5...v4) - 2019-11-11

### Changed

* changed hostname to 'mobro-raspberrypi'
* added MoBro logo to splash screen
* disabled automatic system upgrade as it caused issues
* adapted _config.txt_


## [v4](https://github.com/ModBros/mobro-raspberrypi/compare/v4...v3) - 2019-09-03

### Changed

* minor change to project structure
* tweaks to timeouts and chromium flags

### Fixed

* search for available networks
* handle access point creation failure
* do not write wifi pw to log
* fixed file permissions


## [v3](https://github.com/ModBros/mobro-raspberrypi/compare/v3...v2) - 2019-07-21

### Added


web/util.php
<?php

function getIfNotEof($file, $default): string
{
    return $file && !feof($file) ? trim(fgets($file)) : $default;
}

function closeFile($file)
{
    if ($file) {
        fclose($file);
    }
}

function getFirstLine($filePath, $default): string
{
    $file = fopen($filePath, "r");
    $line = getIfNotEof($file, $default);
    closeFile($file);
    return $line;
}

function getOrDefault($var, $default): string
{
    return trim(!empty($var) || $var == "0" ? $var : $default);
}

function parseProperties($filePath): array
{
    $fileContent = file_get_contents($filePath);
    $result = [];
    $fileContent = str_replace("\r\n", "\n", $fileContent);
    $lines = explode("\n", $fileContent);
    $lastkey = '';
    $appendNextLine = false;
    foreach ($lines as $l) {
        $cleanLine = trim($l);
        if ($cleanLine === '') continue;
        if (strpos($cleanLine, '#') === 0) {
            continue; // is comment ... move on
        }
        $endsWithSlash = substr($l, -1) === '\\';
        if ($appendNextLine) {
            $result[$lastkey] .= "\n" . substr($l, 0, $endsWithSlash ? -1 : 10000);
            if (!$endsWithSlash) { // last line of multi-line property does not end with '\' char
                $appendNextLine = false;
            }
        } else {
            $key = trim(substr($l, 0, strpos($l, '=')));
            $value = substr($l, strpos($l, '=') + 1, $endsWithSlash ? -1 : 10000);
            $lastkey = $key;
            $result[$key] = $value;
            $appendNextLine = $endsWithSlash;
        }
    }
    return $result;
}

function getDriverScripts($dir, $prefix): array
{
    $result = array();
    foreach (scandir($dir) as $key => $value) {
        $full_path = $dir . DIRECTORY_SEPARATOR . $value;
        if (!is_dir($full_path)) {
            if (fnmatch('*show', $value)) {
                if (isset($prefix)) {
                    $result[$full_path] = $prefix . " - " . $value;
                } else {
                    $result[$full_path] = $value;
                }
            }
        }
    }
    return $result;
}

function getGroupedTimeZones(): array
{
    $result = array();
    if (($handle = fopen(Constants::FILE_TIMEZONES, "r")) !== FALSE) {
        while (($data = fgetcsv($handle, 1000, "/")) !== FALSE) {
            if (!array_key_exists($data[0], $result)) {
                $result[$data[0]] = array();
            }
            $result[$data[0]][implode("/", $data)] = end($data);
        }
        fclose($handle);
    }
    return $result;
}

function getGoodTFTDrivers(): array
{
    return getDriverScripts(Constants::DIR_DRIVER_GOODTFT, null);
}

function getWaveshareDrivers(): array
{
    return getDriverScripts(Constants::DIR_DRIVER_WAVESHARE, null);
}

function getOfficialRaspberryPiDrivers(): array
{
    return [
        'pi7' => 'Official 7" touchscreen display'
    ];
}

function getOtherDriverOptions(): array
{
    return [
        'default' => 'Default (HDMI, flat ribbon cable, ...)',
    ];
}

function getAllDrivers(): array
{
    return array_merge(
        getOtherDriverOptions(),
        getOfficialRaspberryPiDrivers(),
        getDriverScripts(Constants::DIR_DRIVER_GOODTFT, 'GoodTFT'),
        getDriverScripts(Constants::DIR_DRIVER_WAVESHARE, 'WaveShare')
    );
}

function getDefaultScreensaverOptions(): array
{
    return [
        'disabled' => 'Disabled',
        'black.html' => 'Black',
    ];
}

function getClockScreensavers(): array
{
    return [
        'clock_date.php' => 'Date + Time',
        'pong.php' => 'Pong',
        'particle_clock.php' => 'Particle Clock (animated)',
        'snow_clock.php' => 'Snow Clock (animated)',
    ];
}

function getMiscScreensavers(): array
{
    return [
        'particle.html' => 'Particles (animated)',
        'snow.html' => 'Snow (animated)',
        'nyancat.html' => 'Nyan Cat 🌈 (animated)',
        'mobro_boot_light.html' => 'MoBro Boot screen (light)',
        'mobro_boot_dark.html' => 'MoBro Boot screen (dark)',
        'mobro_logo_dark.html' => 'MoBro Logo',
        'mobro_logo_dark_bounce.html' => 'MoBro Logo (animated)',
        'modbros_logo_dark.html' => 'ModBros Logo',
        'modbros_logo_dark_bounce.html' => 'ModBros Logo (animated)'
    ];
}

function getOnlineScreensavers(): array
{
    return [
        'custom' => 'Custom URL',
        'unsplash.html' => 'Random Unsplash images',
        'https://www.windy.com' => 'Windy',
        'https://www.flightradar24.com' => 'Flightradar',
        'https://www.marinetraffic.com' => 'Marinetraffic'
    ];
}

function getAllScreensavers(): array
{
    return array_merge(
        getDefaultScreensaverOptions(),
        getClockScreensavers(),
        getOnlineScreensavers(),
        getMiscScreensavers()
    );
}

function getOverClocks(): array
{
    if (isPiZero1()) {
        return array(
            'none' => 'None',
            'high' => 'High: 1050MHz ARM, 450MHz core, 450MHz SDRAM, 6 overvolt',
            'turbo' => 'Turbo: 1100MHz ARM, 500MHz core, 500MHz SDRAM, 6 overvolt'
        );
    }
    if (isPiOne()) {
        return array(
            'none' => 'None',
            'modest' => 'Modest: 800MHz ARM, 250MHz core, 400MHz SDRAM, 0 overvolt',
            'medium' => 'Medium: 900MHz ARM, 250MHz core, 450MHz SDRAM, 2 overvolt',
            'high' => 'High: 950MHz ARM, 250MHz core, 450MHz SDRAM, 6 overvolt',
            'turbo' => 'Turbo: 1000MHz ARM, 500MHz core, 600MHz SDRAM, 6 overvolt'
        );
    }
    if (isPiTwo()) {
        return array(
            'none' => 'None',
            'high' => 'High: 1000MHz ARM, 500MHz core, 500MHz SDRAM, 2 overvolt',
        );
    }
    return array(
        'none' => 'Unavailable'
    );
}

function getSupportedHdmiModesGrouped(): array
{
    return [
        'CEA' => getSupportedModes('CEA', 1),
        'DMT' => getSupportedModes('DMT', 2)
    ];
}

function getSupportedHdmiModes(): array
{
    return array_merge(
        getSupportedModes('CEA', 1),
        getSupportedModes('DMT', 2)
    );
}

function getSupportedModes(string $group, int $prefix): array
{
    $json = shell_exec('/opt/vc/bin/tvservice -m ' . $group . ' -j');
    $dec = json_decode($json);
    $result = array();
    for ($idx = 0; $idx < count($dec); $idx++) {
        $obj = (array)$dec[$idx];
        $scan = $obj['scan'] == 'i' ? 'interlaced' : 'progressive';
        $result[$prefix . ':' . $obj["code"]] = $obj["width"] . 'x' . $obj["height"] . ' @ ' . $obj["rate"] . 'Hz ' . $obj["aspect_ratio"] . ', ' . $scan;
    }
    return $result;
}

function isPiZero1(): bool
{
    return shell_exec('grep -c "^Revision\s*:\s*[ 123][0-9a-fA-F][0-9a-fA-F][0-9a-fA-F]0[9cC][0-9a-fA-F]$" /proc/cpuinfo') > 0;
}

function isPiZero2(): bool
{
    return shell_exec('grep -c "^Raspberry Pi Zero 2 Rev .*$" /proc/device-tree/model') > 0;
}

function isPiOne(): bool
{
    $isPiOne = shell_exec('grep -c "^Revision\s*:\s*00[0-9a-fA-F][0-9a-fA-F]$" /proc/cpuinfo') > 0;
    $isPiOne |= shell_exec('grep -q "^Revision\s*:\s*[ 123][0-9a-fA-F][0-9a-fA-F][0-9a-fA-F]0[0-36][0-9a-fA-F]$" /proc/cpuinfo') > 0;
    return $isPiOne;
}

function isPiTwo(): bool
{
    return shell_exec('grep -c "^Revision\s*:\s*[ 123][0-9a-fA-F][0-9a-fA-F][0-9a-fA-F]04[0-9a-fA-F]$" /proc/cpuinfo') > 0;
}

function isPiFour(): bool
{
    return shell_exec('grep -q "^Revision\s*:\s*[ 123][0-9a-fA-F][0-9a-fA-F]3[0-9a-fA-F][0-9a-fA-F][0-9a-fA-F]$" /proc/cpuinfo') > 0;
}

function overclockSupported(): bool
{
    return isPiZero1() || isPiOne() || isPiTwo();
}
